cmd/kpod/history.go

```diff
@@ -83,226 +94,183 @@ func historyCmd(c *cli.Context) error {
+	img, err := is.Transport.GetStoreImage(store, ref)
 	if err != nil {
-		return nil, errors.Errorf("no such image %q: %v", opts.image, err)
+		return nil, nil, "", errors.Wrapf(err, "no such image %q", image)
 	}
 
 	systemContext := common.GetSystemContext("")
 
 	src, err := ref.NewImage(systemContext)
 	if err != nil {
-		return nil, errors.Errorf("error instantiating image %q: %v", opts.image, err)
+		return nil, nil, "", errors.Wrapf(err, "error instantiating image %q", image)
 	}
 
 	oci, err := src.OCIConfig()
 	if err != nil {
-		return nil, err
+		return nil, nil, "", err
 	}
 
-	history := oci.History
-	layers := src.LayerInfos()
-	count := 1
-	// iterating backwards to get newwest to oldest
+	return oci.History, src.LayerInfos(), img.ID, nil
+}
+
+// getHistorytemplateOutput gets the modified history information to be printed in human readable format
+func getHistoryTemplateOutput(history []v1.History, layers []types.BlobInfo, imageID string, opts historyOptions) (historyOutput []historyTemplateParams) {
+	var (
+		outputSize  string
+		createdTime string
+		createdBy   string
+		count       = 1
+	)
 	for i := len(history) - 1; i >= 0; i-- {
-		if i == len(history)-1 {
-			imageID = img.ID
-		} else {
+		if i != len(history)-1 {
 			imageID = "<missing>"
 		}
+		if !opts.noTrunc && i == len(history)-1 {
+			imageID = imageID[:idTruncLength]
+		}
 
+		var size int64
 		if !history[i].EmptyLayer {
 			size = layers[len(layers)-count].Size
 			count++
-		} else {
-			size = 0
 		}
 
 		if opts.human {
-			createdTime = outputHumanTime(history[i].Created) + " ago"
+			createdTime = units.HumanDuration(time.Since((*history[i].Created))) + " ago"
 			outputSize = units.HumanSize(float64(size))
 		} else {
-			createdTime = outputTime(history[i].Created)
+			createdTime = (history[i].Created).Format(time.RFC3339)
 			outputSize = strconv.FormatInt(size, 10)
 		}
-		params := historyOutputParams{
+
+		createdBy = strings.Join(strings.Fields(history[i].CreatedBy), " ")
+		if !opts.noTrunc && len(createdBy) > createdByTruncLength {
+			createdBy = createdBy[:createdByTruncLength-3] + "..."
+		}
+
+		params := historyTemplateParams{
 			ID:        imageID,
 			Created:   createdTime,
-			CreatedBy: history[i].CreatedBy,
+			CreatedBy: createdBy,
 			Size:      outputSize,
 			Comment:   history[i].Comment,
 		}
-
-		layerAll = append(layerAll, params)
-	}
-
-	output, err := json.MarshalIndent(layerAll, "", "\t\t")
-	if err != nil {
-		return nil, errors.Errorf("error marshalling to JSON: %v", err)
-	}
-
-	if err = src.Close(); err != nil {
-		return nil, err
-	}
-
-	return output, nil
-}
-
-// historyToGeneric makes an empty array of interfaces for output
-func historyToGeneric(params []historyOutputParams) []interface{} {
-	genericParams := make([]interface{}, len(params))
-	for i, v := range params {
-		genericParams[i] = interface{}(v)
+		historyOutput = append(historyOutput, params)
 	}
-	return genericParams
+	return
```

Comment:
I'm beat and am not seeing it, so I'll ask.  When you run 'docker history' you're shown layers that are no longer in play and they're listed as '<missing>'.  How are we handling these layers by ignoring them and not printing them?  Or printing them as though they're not '<missing>'.  I'm not seeing <missing> here, is it just part of the layer or am I missing it?

Reply:
@TomSweeneyRedHat I actually never came across an image with layers that are no longer in play. Could you give me an image that has this and I can test it and see for myself. I just access the `History` struct in the package v1, and the `BlobInfo` for the size of each layer.